Add .x as XDR/RPCGEN

dylex opened the pull request:
XDR/RPC language as documented in RFC5531, RFC4506.

Also adding random Logos.x samples to disambiguate, taken from random github search results -- hopefully this is okay license-wise.

Vim has [syntax highlighting for rpcgen](http://ftp.vim.org/vim/runtime/syntax/rpcgen.vim) but I haven't found any other syntax highlighting.  I can write one if necessary.

Many in-the-wild [examples on github](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ax+struct&type=Code&ref=searchresults).

lildude commented:
@dylex Can you please provide references to the samples you've included. There's no indication within them what their license is.

dylex commented:
I presume you mean for the Logos samples?  The RPC samples should all be documented I think.  I've added links to their original source, though the licenses are not necessarily clear there either.  I don't really what Logos is or what the included samples are for, but they seem to be the right language.  My goal was just to provide sufficient disambiguation for RPC.

lildude commented:
Thanks @dylex.  Unfortunately, we can't include any sample files if it's not clear exactly what license the code has been released under.

dylex commented:
Understood.  Feel free to take only the first commit aad8ab9f3138c247dc8b2fffbb1e92239c8a6214 that adds RPC in that case, since it does not contain any questionably licensed files.

lildude commented:
Ooops, completely forgot about this.  Thanks for the update.

If you're still interested in this, we're going to need:

- a Logos sample file else all those `.x` files will now become identified as RPC - https://github.com/JonasGessner/NoCarrier/blob/master/NoCarrier.x looks good as it's clearly MIT licensed
- a heuristic to disambiguate between the two languages as you originally provided
- info on where the samples have come from and the license they're licensed under - this doesn't need to be in the files themselves; a comment in this PR is sufficient.

Can you please also remove the nfs.x file - it's quite large and contains a lot of duplication in the "structure" of the content.

dylex commented:
The commit messages should contain all that information already.

lildude commented:
Please state it in the PR as a comment. Our legal peeps would prefer it if such info is easy to find. Commits are also squashed when the PR is merged.

dylex commented:
Sample licenses and sources:

https://github.com/JonasGessner/NoCarrier/blob/master/NoCarrier.x - MIT
https://github.com/basicthinker/PTMC/blob/cf31f4e46654cf967d2b3731f16fdb570be26ba0/llvm-gcc-R3/gcc/testsuite/objc/execute/string1.x - GPL2
https://github.com/fantastxmw/MobileShadowSocks/blob/f6415578fa00e607491c01106be83bba99008e1b/perapp-plugin/Tweak.x - GPL3
https://github.com/hbang/DailyPaper/blob/d1b3e838889c9e4754116fc5bf07fdc370d893b2/NCHax.x - Apache
rpc.x - RFC5531
rusers.x, yp.x - glibc, MIT3

Alhadis commented on lib/linguist/heuristics.rb:
```diff
@@ -513,5 +513,13 @@ def call(data)
     end
   
+    disambiguate ".x" do |data|
+      if /\<(program|version)\s+\w+\s*{|\<union\s+\w+\s+switch\s*\(/.match(data)
```
Why is `\<` escaped?

Alhadis commented on lib/linguist/languages.yml:
```diff
@@ -3752,6 +3752,19 @@ RMarkdown:
+  - ".x"
+  interpreters:
+  - rpcgen
+  tm_scope: none
```
> *Vim has syntax highlighting for rpcgen but I haven't found any other syntax highlighting. I can write one if necessary.*

You could use Objective-C's grammar for highlighting, as the language appears to be similar enough in syntax. Otherwise, a `tm_scope: none` will cause all RPC files to be shown on GitHub without colour.

dylex commented:
Oops, `\<` is a pcre-ism that I forgot doesn't exist in ruby.  Changed to `\b`.

Changed `tm_scope` to `source.c` since it's C-like (or at least C-based).

pchaigno commented:
@dylex Are [these files](https://github.com/search?p=99&q=extension%3Ax+NOT+struct&type=Code&utf8=%E2%9C%93) RPC as well?

pchaigno commented on samples/RPC/rusers.x:
```diff
@@ -0,0 +1,228 @@
+} = 100002;
+
+#ifdef RPC_HDR
+%
```
All following lines start with `%`. Does that indicate a comment line?

dylex replied:
No, any `%` lines are passed through directly to the generated C file by rpcgen, but that does mean that the lines are not in the RPC language.  Since many RPC files include such inline C code, it seemed to make sense to leave one in a sample. (More [RPC docs here]( https://docs.oracle.com/cd/E19683-01/816-1435/rpcgenpguide-21470/index.html).)

pchaigno replied:
> Since many RPC files include such inline C code, it seemed to make sense to leave one in a sample.

👍

dylex commented:
@pchaigno hm, no those look like ld linker scripts, which lingust already has registered as `.ld`.

pchaigno commented on lib/linguist/languages.yml:
```diff
@@ -3753,6 +3753,19 @@ RMarkdown:
+  - ".x"
+  interpreters:
+  - rpcgen
+  tm_scope: source.c
```
I found a [grammar for RPC on GitHub](https://github.com/shakeelmohamed/sublime-rpcgen). Could you check if it does a better job than the C grammar, especially regarding the `%` lines mentioned previously? You can use [Lightshow](https://github-lightshow.herokuapp.com/) to test it without installing Sublime Text.

dylex replied:
It looks pretty much just like C, and doesn't seem to add much useful (like `program` or `string` keywords or special `%` processing). I'd say it's probably not worth it.

pchaigno replied:
Ok. Thanks for checking.

pchaigno commented on lib/linguist/languages.yml:
```diff
@@ -3753,6 +3753,19 @@ RMarkdown:
+  extensions:
+  - ".x"
+  interpreters:
+  - rpcgen
```
Are there actually RPC files using a `rpcgen` shebang? I [couldn't find any with GitHub searches](https://github.com/search?p=1&q=%22bin%2Frpcgen%22&type=Code&utf8=%E2%9C%93), but it's not particularly easy to search for strings with special characters.

dylex replied:
Not that I've ever seen, but rpcgen is the only program I know of that could.

Actually, looks like my rpcgen gets confused about the `#!` line anyway, so perhaps there's no point.

pchaigno replied:
Hm. Yes, it's probably better to remove it if it's not used.

lildude commented:
@pchaigno are you happy with this PR now?

pchaigno commented:
I think we should also add the `.x` file extension to [the Linker Script entry](https://github.com/github/linguist/blob/8da6ddf9d97ee1cd1d7119f3e8a5249df7d1590a/lib/linguist/languages.yml#L2362-L2371). But since these files are already misclassified as Logo, maybe we can defer this to a subsequent PR...?

lildude commented:
> I think we should also add the .x file extension to the Linker Script entry. But since these files are already misclassified as Logo, maybe we can defer this to a subsequent PR...?

Yeah, I think we can defer this to a subsequent PR as a quick bit of digging suggests there are several other `.x` variants for linker scripts according to the comments at https://github.com/pathscale/binutils/blob/5c2c133020e41fc4aadd80a99156d2cea4754b96/ld/genscripts.sh#L55-L66
